File: blockchain.py
import hashlib
import time

class Block:
    def __init__(self, index, timestamp, file_name, file_hash, previous_hash):
        self.index = index
        self.timestamp = timestamp
        self.file_name = file_name
        self.file_hash = file_hash
        self.previous_hash = previous_hash
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        data = f"{self.index}{self.timestamp}{self.file_name}{self.file_hash}{self.previous_hash}"
        return hashlib.sha256(data.encode()).hexdigest()

    def to_dict(self):
        return {
            "index": self.index,
            "timestamp": self.timestamp,
            "file_name": self.file_name,
            "file_hash": self.file_hash,
            "previous_hash": self.previous_hash,
            "hash": self.hash,
        }

class Blockchain:
    def __init__(self):
        self.chain = [self.create_genesis_block()]

    def create_genesis_block(self):
        return Block(0, time.time(), "Genesis Block", "0", "0")

    def add_block(self, file_name, file_hash):
        previous_block = self.chain[-1]
        new_block = Block(
            len(self.chain), 
            time.time(), 
            file_name, 
            file_hash, 
            previous_block.hash
        )
        self.chain.append(new_block)

    def is_chain_valid(self):
        for i in range(1, len(self.chain)):
            current_block = self.chain[i]
            previous_block = self.chain[i - 1]

            if current_block.hash != current_block.calculate_hash():
                return False

            if current_block.previous_hash != previous_block.hash:
                return False

        return True

    def to_dict(self):
        return [block.to_dict() for block in self.chain]

File: blockchain.go
package main

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"
)

type Block struct {
	Index        int    `json:"index"`
	Timestamp    string `json:"timestamp"`
	FileName     string `json:"file_name"`
	FileHash     string `json:"file_hash"`
	PreviousHash string `json:"previous_hash"`
	Hash         string `json:"hash"`
}

type Blockchain struct {
	Blocks []Block
	mu     sync.Mutex
}

var blockchain = Blockchain{
	Blocks: []Block{createGenesisBlock()},
}

func createGenesisBlock() Block {
	return Block{
		Index:        0,
		Timestamp:    time.Now().String(),
		FileName:     "Genesis Block",
		FileHash:     "",
		PreviousHash: "0",
		Hash:         calculateHash(0, time.Now().String(), "Genesis Block", "", "0"),
	}
}

func calculateHash(index int, timestamp, fileName, fileHash, previousHash string) string {
	record := fmt.Sprintf("%d%s%s%s%s", index, timestamp, fileName, fileHash, previousHash)
	h := sha256.New()
	h.Write([]byte(record))
	return fmt.Sprintf("%x", h.Sum(nil))
}

func (bc *Blockchain) AddBlock(fileName, fileHash string) {
	bc.mu.Lock()
	defer bc.mu.Unlock()

	lastBlock := bc.Blocks[len(bc.Blocks)-1]
	newBlock := Block{
		Index:        len(bc.Blocks),
		Timestamp:    time.Now().String(),
		FileName:     fileName,
		FileHash:     fileHash,
		PreviousHash: lastBlock.Hash,
		Hash:         calculateHash(len(bc.Blocks), time.Now().String(), fileName, fileHash, lastBlock.Hash),
	}
	bc.Blocks = append(bc.Blocks, newBlock)
}

func (bc *Blockchain) GetBlockchain() []Block {
	bc.mu.Lock()
	defer bc.mu.Unlock()
	return bc.Blocks
}

func (bc *Blockchain) Validate() bool {
	bc.mu.Lock()
	defer bc.mu.Unlock()

	for i := 1; i < len(bc.Blocks); i++ {
		currentBlock := bc.Blocks[i]
		previousBlock := bc.Blocks[i-1]

		if currentBlock.Hash != calculateHash(currentBlock.Index, currentBlock.Timestamp, currentBlock.FileName, currentBlock.FileHash, currentBlock.PreviousHash) {
			return false
		}

		if currentBlock.PreviousHash != previousBlock.Hash {
			return false
		}
	}
	return true
}

func handleAddBlock(w http.ResponseWriter, r *http.Request) {
	var data struct {
		FileName string `json:"file_name"`
		FileHash string `json:"file_hash"`
	}
	if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}
	blockchain.AddBlock(data.FileName, data.FileHash)
	w.WriteHeader(http.StatusCreated)
}

func handleGetBlockchain(w http.ResponseWriter, r *http.Request) {
	json.NewEncoder(w).Encode(blockchain.GetBlockchain())
}

func handleValidate(w http.ResponseWriter, r *http.Request) {
	if blockchain.Validate() {
		w.Write([]byte("Blockchain is valid"))
	} else {
		w.Write([]byte("Blockchain is invalid"))
	}
}

func main() {
	http.HandleFunc("/add_block", handleAddBlock)
	http.HandleFunc("/blockchain", handleGetBlockchain)
	http.HandleFunc("/validate", handleValidate)

	fmt.Println("Blockchain server running on port 8080...")
	http.ListenAndServe(":8080", nil)
}
